dds/idl/typeobject_generator.cpp

```diff
@@ -1190,251 +1731,630 @@ typeobject_generator::strong_connect(AST_Type* type, const std::string& anonymou
-          }
-        }
-        member.common.label_seq.sort();
+    if (be_global->is_external(field)) {
+      minimal_member.common.member_flags |= OpenDDS::XTypes::IS_EXTERNAL;
+    }
+
+    minimal_member.common.member_type_id = get_minimal_type_identifier(field->field_type());
+    OpenDDS::XTypes::hash_member_name(minimal_member.detail.name_hash, field->local_name()->get_string());
+    minimal_to.minimal.struct_type.member_seq.append(minimal_member);
+
+    OpenDDS::XTypes::CompleteStructMember complete_member;
+    complete_member.common.member_id = minimal_member.common.member_id;
+    complete_member.common.member_flags = minimal_member.common.member_flags;
+    complete_member.common.member_type_id = get_complete_type_identifier(field->field_type());
+
+    complete_member.detail.name = field->local_name()->get_string();
+    set_builtin_member_annotations(field, complete_member.detail.ann_builtin);
 
-        OpenDDS::XTypes::hash_member_name(member.detail.name_hash, branch->local_name()->get_string());
+    complete_to.complete.struct_type.member_seq.append(complete_member);
+  }
+
+  minimal_to.minimal.struct_type.member_seq.sort();
+  complete_to.complete.struct_type.member_seq.sort();
+
+  update_maps(type, minimal_to, complete_to);
+}
+
+void
+typeobject_generator::generate_union_type_identifier(AST_Type* type)
+{
+  AST_Union* const n = dynamic_cast<AST_Union*>(type);
+  AST_Type* discriminator = n->disc_type();
+  const Fields fields(n);
+
+  const ExtensibilityKind exten = be_global->extensibility(n);
+  const AutoidKind auto_id = be_global->autoid(n);
 
-        to.minimal.union_type.member_seq.append(member);
+  OpenDDS::XTypes::TypeObject minimal_to, complete_to;
+  minimal_to.kind = OpenDDS::XTypes::EK_MINIMAL;
+  minimal_to.minimal.kind = OpenDDS::XTypes::TK_UNION;
+  minimal_to.minimal.union_type.union_flags = extensibility_to_type_flag(exten);
+
+  if (be_global->is_nested(n)) {
+    minimal_to.minimal.union_type.union_flags |= OpenDDS::XTypes::IS_NESTED;
+  }
+
+  if (auto_id == autoidkind_hash) {
+    minimal_to.minimal.union_type.union_flags |= OpenDDS::XTypes::IS_AUTOID_HASH;
+  }
+
+  const TryConstructFailAction trycon = be_global->union_discriminator_try_construct(n);
+  minimal_to.minimal.union_type.discriminator.common.member_flags = try_construct_to_member_flag(trycon);
+  if (be_global->union_discriminator_is_key(n)) {
+    minimal_to.minimal.union_type.discriminator.common.member_flags |= OpenDDS::XTypes::IS_KEY;
+  }
+  minimal_to.minimal.union_type.discriminator.common.type_id = get_minimal_type_identifier(discriminator);
+
+  complete_to.kind = OpenDDS::XTypes::EK_COMPLETE;
+  complete_to.complete.kind = OpenDDS::XTypes::TK_UNION;
+  complete_to.complete.union_type.union_flags = minimal_to.minimal.union_type.union_flags;
+
+  const char* name = type->name()->get_string_copy();
+  complete_to.complete.union_type.header.detail.type_name = name;
+  delete[] name;
+  name = 0;
```

Comment:
Use `scoped`? I also think it should be passed `EscapeContext_StripEscapes` to get the normal IDL name back if this is supposed to be a generic name. Otherwise it's the C++ name. Apply to the other places that do this.